wgengine/magicsock/magicsock.go

```diff
@@ -1665,12 +1683,12 @@ func (c *Conn) receiveIPv4(b []byte) (n int, ep conn.Endpoint, err error) {
 //
 // ok is whether this read should be reported up to wireguard-go (our
 // caller).
-func (c *Conn) receiveIP(b []byte, ipp netip.AddrPort, cache *ippEndpointCache) (ep *endpoint, ok bool) {
+func (c *Conn) receiveIP(b []byte, ipp netip.AddrPort, cache *ippEndpointCache, checkDisco bool) (ep *endpoint, ok bool) {
 	if stun.Is(b) {
 		c.stunReceiveFunc.Load()(b, ipp)
 		return nil, false
 	}
-	if c.handleDiscoMessage(b, ipp, key.NodePublic{}) {
+	if checkDisco && c.handleDiscoMessage(b, ipp, key.NodePublic{}) {
```

Comment:
You're assuming that just because we listened that the BPF is working? I'd want to send a test packet to ourselves and test that we get it first before turning off our UDP disco processing here.

Also, don't we still want to look to see if it's disco here and discard it rather than letting it fall through to wireguard?

Reply:
The latter: yes, noticed that and fixed.

The former: no, I don't think it's necessary to self-test. The combination of ListenPacket and SetBPF succeeding is sufficient, IMO.

Reply:
I repent: I implemented the self-test because it was easy enough to do, and it revealed that the IPv6 BPF program is broken and doesn't appear to receive disco traffic. So, um, debugging that now.

Reply:
hah, lolsobyay?

Reply:
All glory to @andrew-d for figuring it out: when you create a raw UDP socket, the bytes the BPF program receives as input vary based on address family: for a UDPv4 raw socket, it's the entire IP+UDP+payload. For UDPv6, it's just UDP+payload. So obviously, our programs that assumed IP+UDP+payload didn't work.

Cursed, but true. And now fixed. On the plus side, I have empirical proof that the self-test and fallback work correctly :zany_face: